File: algorithm/444.sequence-reconstruction.py
import collections
class Solution(object):
    def sequenceReconstruction(self, org, seqs):
        """
        :type org: List[int]
        :type seqs: List[List[int]]
        :rtype: bool
        """
        
        graph = collections.defaultdict(list)
        indegree_map = {}

        # build graph and count indegree
        for pair in seqs:
            if len(pair) == 0:
                continue

            for i in range(len(pair)):
                node = pair[i]
                _ = graph[node]

                if i + 1 >= len(pair):
                    continue

                neighbor = pair[i + 1]
                _ = graph[neighbor]
                graph[node].append(neighbor)

                if neighbor in indegree_map:
                    indegree_map[neighbor] = indegree_map[neighbor] + 1
                else:
                    indegree_map[neighbor] = 1

        # find node with in degree = 0
        queue = []
        for node in graph:
            if node not in indegree_map:
                queue.append(node)

        # get topological sort with bfs
        result = []
        while(len(queue) > 0):
            if len(queue) > 1:
                return False
            
            # if non-recursive dfs, using queue.pop()
            # reference: https://stackoverflow.com/a/5278667/1389917
            node = queue.pop(0) 
            result.append(node)

            for neighbor in graph[node]:
                if neighbor in indegree_map:
                    indegree_map[neighbor] = indegree_map[neighbor] - 1
                    if indegree_map[neighbor] == 0:
                        queue.append(neighbor)
                        del indegree_map[neighbor]

        if len(indegree_map) > 0:
            return False

        if len(result) != len(org):
            return False

        for i in range(len(result)):
            if result[i] != org[i]:
                return False

        return True

File: leetcode/444.sequence-reconstruction.cpp
class Solution {
public:
    bool sequenceReconstruction(vector<int>& org, vector<vector<int>>& seqs) {
        unordered_map<int, vector<int>> graph;
        unordered_map<int, int> indegree;
        unordered_set<int> nodes;
        
        for (auto i = 0; i < seqs.size(); ++i)
        {
            vector<int> sequence = seqs[i];
            for (auto j = 0; j < sequence.size(); ++j)
            {
                nodes.insert(sequence[j]);
                if (j > 0)
                {
                    indegree[sequence[j]] += 1;
                    graph[sequence[j - 1]].push_back(sequence[j]);
                }
            }
        }
        
        queue<int> q;
        vector<int> res;
        for (auto n : nodes)
        {
            if (indegree.count(n) == 0)
            {
                q.push(n);
            }
        }
        
        while (!q.empty()) {
            
            if (q.size() > 1)
            {
                return false;
            }
            
            int node = q.front();
            q.pop();
            res.push_back(node);
            
            vector<int> neighbors = graph[node];
            for (auto i = 0; i < neighbors.size(); ++i)
            {
                indegree[neighbors[i]] -= 1;
                if (indegree[neighbors[i]] == 0)
                {
                    indegree.erase(neighbors[i]);
                    q.push(neighbors[i]);
                }
            }
        }
        
        return indegree.empty() ? res == org: false;
    }
};
